expression/builtin_string.go

```diff
@@ -200,8 +200,12 @@ func (c *concatFunctionClass) getFunction(args []Expression, ctx context.Context
 		return nil, errors.Trace(err)
 	}
 	for i := range args {
-		bf.tp.Flag |= mysql.BinaryFlag
-		bf.tp.Flen += args[i].GetType().Flen
+		argType := args[i].GetType()
+		if types.IsBinaryStr(argType) {
+			bf.tp.Flag |= mysql.BinaryFlag
```

Comment:
use `SetBinChsClnFlag(...)` to set flag/charset/collation to binary and break the execution of `for` loop